From e0bb62c708d3435f49645d0f3d6ce871174c1cc1 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:36:45 -0400
Subject: [virt] kvm/eventfd: switch to fdget

Message-id: <1406040016-3289-11-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86089
O-Subject: [RHEL7 PATCH v2 010/221] kvm eventfd: switch to fdget
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Al Viro <viro@zeniv.linux.org.uk>

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
(cherry picked from commit cffe78d92c217a57f57ec6743f71adfe39ea543e)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/virt/kvm/eventfd.c b/virt/kvm/eventfd.c
index 1550637..abe4d60 100644
--- a/virt/kvm/eventfd.c
+++ b/virt/kvm/eventfd.c
@@ -291,7 +291,7 @@ kvm_irqfd_assign(struct kvm *kvm, struct kvm_irqfd *args)
 {
  struct kvm_irq_routing_table *irq_rt;
  struct _irqfd *irqfd, *tmp;
- struct file *file = NULL;
+ struct fd f;
  struct eventfd_ctx *eventfd = NULL, *resamplefd = NULL;
  int ret;
  unsigned int events;
@@ -306,13 +306,13 @@ kvm_irqfd_assign(struct kvm *kvm, struct kvm_irqfd *args)
  INIT_WORK(&irqfd->inject, irqfd_inject);
  INIT_WORK(&irqfd->shutdown, irqfd_shutdown);
 
- file = eventfd_fget(args->fd);
- if (IS_ERR(file)) {
-  ret = PTR_ERR(file);
-  goto fail;
+ f = fdget(args->fd);
+ if (!f.file) {
+  ret = -EBADF;
+  goto out;
  }
 
- eventfd = eventfd_ctx_fileget(file);
+ eventfd = eventfd_ctx_fileget(f.file);
  if (IS_ERR(eventfd)) {
   ret = PTR_ERR(eventfd);
   goto fail;
@@ -391,7 +391,7 @@ kvm_irqfd_assign(struct kvm *kvm, struct kvm_irqfd *args)
         lockdep_is_held(&kvm->irqfds.lock));
  irqfd_update(kvm, irqfd, irq_rt);
 
- events = file->f_op->poll(file, &irqfd->pt);
+ events = f.file->f_op->poll(f.file, &irqfd->pt);
 
  list_add_tail(&irqfd->list, &kvm->irqfds.items);
 
@@ -408,7 +408,7 @@ kvm_irqfd_assign(struct kvm *kvm, struct kvm_irqfd *args)
   * do not drop the file until the irqfd is fully initialized, otherwise
   * we might race against the POLLHUP
   */
- fput(file);
+ fdput(f);
 
  return 0;
 
@@ -422,9 +422,9 @@ fail:
  if (eventfd && !IS_ERR(eventfd))
   eventfd_ctx_put(eventfd);
 
- if (!IS_ERR(file))
-  fput(file);
+ fdput(f);
 
+out:
  kfree(irqfd);
  return ret;
 }
-- 
1.7.1